scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -1200,6 +1225,56 @@ static bool keep_running(struct cpu_ctx *cctx, struct task_struct *p)
 	return false;
 }
 
+/*
+ * Helper function to consume layer DSQs when a system is under starvation.
+ */
+static __always_inline bool consume_starved(s32 cpu)
+{
+	struct layer *layer;
+	u32 idx;
+	u64 decay_vtime_sum = 0;
+	u64 layer_weight_sum = 0;
+	u64 decay_proportion_vtime = 0;
+	u64 vtime_weighted_decay = 0;
+
+	bpf_for(idx, 0, nr_layers) {
+		layer = MEMBER_VPTR(layers, [idx]);
+		decay_vtime_sum += layer->vtime_decay;
+		layer_weight_sum += layer->weight;
+	}
+
+	// Use a per cpu layer to consume from. With enough CPUs this
+	// should consume starved layers efficiently.
+	// TODO: Make this work for nr_layers > nr_possible_cpus
+	idx = ((u32)cpu & 0xfff) % nr_layers;
+	if (idx > nr_layers)
+		return false;
+
+	layer = MEMBER_VPTR(layers, [idx]);
+	if (!layer)
+		return false;
+
+	if (layer_weight_sum > decay_vtime_sum)
+		return false;
+
+	/*
+	 * If the weighted decayed vtime of the layer is less than the
+	 * proportional weighted vtime then the layer is potentially starved
+	 * and should be consumed from.
+	 */
+	vtime_weighted_decay = decay_vtime_sum / layer_weight_sum;
```

Comment:
I don't think this calculation is correct. Imagine three layers with weight 1, 2 and 100 with the third layer completely idle and the first two layers competing against each other. The expected result is 1:2 distribution but with the above it will get 101:102 distribution.